FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+#include "FEXCore/Utils/LogManager.h"
+#include "Interface/IR/IR.h"
+#include "Interface/IR/IREmitter.h"
+#include "Interface/IR/PassManager.h"
+#include <FEXCore/IR/IR.h>
+#include <FEXCore/Utils/Profiler.h>
+#include <FEXCore/fextl/deque.h>
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <stdint.h>
+
+namespace FEXCore::IR {
+
+// FixedSizeStack is a model of the x87 Stack where each element in this
+// fixed size stack lives at an offset from top. The top of the stack is at
+// index 0.
+template<typename T, size_t MaxSize>
+class FixedSizeStack {
+public:
+  enum class Slot { UNUSED, INVALID, VALID };
+  FixedSizeStack() {
+    buffer.resize(MaxSize, {Slot::UNUSED, T()});
+  }
+
+  void push(const T& value) {
+    rotate();
+    buffer.front() = {Slot::VALID, value};
+  }
+
+  // Rotate the elements with the direction controlled by Right
+  void rotate(bool Right = true) {
+    if (Right) {
+      // Right rotation
+      std::pair<Slot, T> temp = buffer.back();
+      buffer.pop_back();
+      buffer.push_front(temp);
+      TopOffset++;
+    } else {
+      // Left rotation
+      std::pair<Slot, T> temp = buffer.front();
+      buffer.pop_front();
+      buffer.push_back(temp);
+      TopOffset--;
+    }
+  }
+
+  void pop() {
+    buffer.front() = {Slot::INVALID, T()};
+    rotate(false);
+  }
+
+  const std::pair<Slot, T>& top(size_t Offset = 0) const {
+    return buffer[Offset];
+  }
+
+  void setTop(T Value, size_t Offset = 0) {
+    buffer[Offset] = {Slot::VALID, Value};
+  }
+
+  bool isValid(size_t Offset) const {
+    return buffer[Offset].first;
+  }
+
+  inline void clear() {
+    for (auto& element : buffer) {
+      element = {Slot::UNUSED, T()}; // Set all elements as invalid
+    }
+    TopOffset = 0;
+  }
+
+  void dump() const {
+    for (size_t i = 0; i < MaxSize; i++) {
+      const auto& [Valid, Element] = buffer[i];
+      if (Valid == Slot::VALID) {
+        LogMan::Msg::DFmt("ST{}: 0x{:x}", i, (uintptr_t)(Element.StackDataNode));
+      }
+    }
+  }
+
+  constexpr size_t size() const {
+    return MaxSize;
+  }
+
+  int8_t getTopOffset() const {
+    return TopOffset;
+  }
+
+  void setTopOffset(int8_t Offset) {
```

Comment:
This getter/setter pair doesn't seem to add any value instead of just making TopOffset public. Could we drop them?